README.md
**Pre-alpha.** This project is still under development and not ready to be used yet. Currently under GPL3. See LICENSE.txt.

# ImageFilter

This browser extension filters images on webpages you visit to improve visibility and detail. It uses SVG filters and allows users to edit their source directly. Features include histogram generation for equalization, dynamically adjustable values with keyboard shortcuts and persistent per-site settings. Images which are classified as having text content or having relatively few colours can be ignored.

## Browser Support

- Chrome
- Firefox

    Very limited - popup and options page don't work. If installed via WebExtensions when it is released rather than addon SDK it may simply work.

## Quick Start

1. Install the extension via the chrome webstore or using a download from the releases section here.
2. Navigate to a page with images.
3. Click on the extension's popup button ![popup img][popup] to open a small popup with options.
4. Here you may enable/disable filtering or choose which filter you want to apply. The filters are configurable at the options page (linked from the popup).

Note there are two sections in the popup: one set for the default behaviour which applies to all sites, and another for per-site overrides. The overrides have a green border when active and can be cleared to allow the default options to apply again. These settings are persistent (also synced on your google account for chrome).

## How it works

ImageFilter has four components.

1. `imagefinder.js` which uses mutation observers to detect all images added to the page.

	This works dynamically so the pages such as youtube which entirely use AJAX to change tabs don't break the plugin and require a reload.

	It also detects images added via CSS, including inline `style` attributes `style` tags and linked `style` files. Again these are supported dynamically and even changes to image `src` attributes and `background-image` properties are handled.

2. `filtermanager.js` is the main controlling component which handles callbacks from `imagefinder.js`, decides whether to filter and keeps a record of what is actually filtered.

3. `histogram.js` builds histograms from images to be used in image classification and in the actual SVG filters.

4. `filter.js` does the actual filtering. It takes a histogram object, creates an SVG object with a `<filter>` tag and a CSS class to apply the SVG filter. Images can be filtered by appending the CSS class.

The following files handle user input and options.

- `content.js` is the entry point, which creates an `FilterManager` from `filtermanager.js`.
- `popup.html/js` is the ![popup img][popup] popup which provide quick access to changing default and per-site enabled/disable and filter choice options. It includes sliders for custom variables and a link to the options page.
- `options.html/js` the less common options are found here, including shortcuts and filter source code. It includes examples of the syntax for integrating histogram and custom variables with the SVG filters, which for convenience is duplicated below.

For chrome and firefox WebExtensions, `manifest.json` contains the extension metadata.

For firefox addon SDK support, `package.json` and `index.js` attempt to wrap and replicate functionality in the chrome code. `storage.js`, `messaging.js` and `crossbrowser.js` contain browser-specific code to support this.

Any configurable variables are found at the top of each file.

## Custom SVG Filters

The extension injects `<svg>` and `<filter>` tags, leaving the contents of `<filter>` for the user to define, although a number of default filters are provided.
Documentation for the different filter effects can be found at the
[W3C SVG Filter Effects page](https://www.w3.org/TR/SVG/filters.html).

Filters may use custom values, named `V1`, `V2` and `V3` to allow parameters to be adjusted dynamically.
Use `{{ V1 }}`, `{{ V2 }}` and `{{ V3 }}` to have it substituted with their value between
zero and one. Javascript equations within the braces are evaluated and
initial values can be set by including xml comments with `V=` assignments, for example below.

	<!-- V1=0.5 -->
	<feComponentTransfer>
	  <feFuncR type="linear" slope="{{V1*4-1}}" intercept="{{-0.5 * (V1*4-1) + 0.5}}"/>
	  <feFuncG type="linear" slope="{{V1*4-1}}" intercept="{{-0.5 * (V1*4-1) + 0.5}}"/>
	  <feFuncB type="linear" slope="{{V1*4-1}}" intercept="{{-0.5 * (V1*4-1) + 0.5}}"/>
	</feComponentTransfer>

Histograms are generated for all images, although sometimes this isn't possible.
If one is available, its values can be injected into a filter with
`%HR`, `%HG` and `%HB`
for red, green and blue channels respectively. `%HY` inserts luminance.
These values are the cumulative sum of the raw histogram, 255 values separated by spaces.
For efficiency, the number of values can be reduced, for `%16HY` enters
just 16 values from the cumulative sum of the luminance histogram. A full example is below,
filtering images by luminance.

	<feComponentTransfer>
	  <feFuncR type="table" tableValues="%16HY"/>
	  <feFuncG type="table" tableValues="%16HY"/>
	  <feFuncB type="table" tableValues="%16HY"/>
	</feComponentTransfer>

[popup]:https://raw.githubusercontent.com/pknowles/ImageFilter/master/icon16.png

# License

GPL3. See LICENSE.txt.

Dependencies:

- JQuery (MIT)
- Mousetrap.js (Apache 2.0)


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.


### Added
- Minimum size for image filtering in the options page. 


### Changed
-Changed it so that filter variants (using the slider bar and keystrokes) can remain through different pages and reloads. Similar to what the old Firefox version does. Chrome only for the moment.
- Bugfixes related to the popup and the Firefox port.
- Various tweaks to make the Web Extension work on Firefox. 

## [0.1.1] - 31/1/2018

imagefilter.js

//FIXME: bug when changing image source of a parent node filtered after removing a child
//TODO: pause histogram generaiton when disabled
//TODO: clean up filters when they're not needed
//TODO: faster isPicture choice with css?: https://stackoverflow.com/questions/2481414/how-do-i-select-a-div-with-class-a-but-not-with-class-b

function ImageFilterer() {
	this.animationUpdateFrequency = 1000;
	//this.golayMeritThreshold = 0.08;

	this.finder = new ImageFinder();
	this.images = [];
	this.uidNext = 0;
	this.histograms = {};
	this.animatedHistograms = {};
	this.svgFilters = {};
	this.onlyPictures = true;
	this.inverted = true;

	this.customValueCache = {V1:0.5, V2:0.5, V3:0.5};

	var that = this;
	this.start = this.finder.start.bind(this.finder);
	this.finder.sourceAdded = this.sourceAdded.bind(this);
	this.finder.sourceRemoved = this.sourceRemoved.bind(this);
	this.finder.imageAdded = this.imageAdded.bind(this);
	this.finder.imageRemoved = this.imageRemoved.bind(this);

	this.nextOverrideID = 0;
	this.defaultFilter = null;
	this.filters = {};
	this.filterOverrides = {};
	this.filterSources = [
		'<feComponentTransfer in="SourceGraphic" result="Current"><feFuncR type="table" tableValues="%HR"/> <feFuncG type="table" tableValues="%HG"/> <feFuncB type="table" tableValues="%HB"/></feComponentTransfer>'
		+ '<feComponentTransfer in="SourceGraphic" result="Last"><feFuncR type="table" tableValues="%LHR"/> <feFuncG type="table" tableValues="%LHG"/> <feFuncB type="table" tableValues="%LHB"/> </feComponentTransfer>'
		+ '<feComposite in="Last" in2="Current" operator="arithmetic" k1="0" k2="1" k3="0" k4="0"> <animate attributeName="k2" from="1" to="0" dur="1s" /> <animate attributeName="k3" from="0" to="1" dur="1s" /> </feComposite>'
		,
		'<feComponentTransfer in="SourceGraphic" result="A"> <feFuncR type="table" tableValues="%HR"/> <feFuncG type="table" tableValues="%HG"/> <feFuncB type="table" tableValues="%HB"/> </feComponentTransfer>'
		,
		'<feColorMatrix in="SourceGraphic" type="matrix" values="-1 0 0 0 1 0 -1 0 0 1 0 0 -1 0 1 0 0 0 1 0"/>'
	];
}

//an element to put debug info into the document
ImageFilterer.debugInfo = null;
ImageFilterer.enableDebug = false;

ImageFilterer.prototype.isPicture = function(image, histogram) {
	var w = $(image).width();
	var h = $(image).height();
	if (w * h < 128 * 128)
		return false;

	if (histogram.success)
	{
		//var merit = histogram.getGolayMerit();
		//if (merit < this.golayMeritThreshold)
		//	return false;
		var peakArea = histogram.getPeakArea(0.02);
		if (peakArea > 0.3)
			return false;
	}

	return true;
};

ImageFilterer.prototype.invertAll = function(enabled) {
	/*
	var style = $('#imagefilter-invertall');
	if (!style.length)
	{
		style = $('<style id="imagefilter-invertall"></style>');
		$(document.body).append(style);
	}
	style.html(enable ? '.imagefilter-all {-moz-filter: invert(100%); -webkit-filter: invert(100%); filter: invert(100%);}' : '');
	*/
	this.inverted = enabled;
	if (this.defaultFilter)
		this.defaultFilter.invert(enabled);
	for (var k in this.filters)
		this.filters[k].invert(enabled);
};

ImageFilterer.prototype.setOnlyPictures = function(enabled) {
	this.onlyPictures = enabled;
	//update filter class for every image currently being filtered
	for (var i = 0; i < this.images.length; ++i)
	{
		var image = $(this.images[i]);

		//don't filter images with filtered children
		if (image.data('imagefilter-haschild'))
			continue;

		var filteredClass = image.attr('data-imagefilter-class');
		var apply = true;
		if (this.onlyPictures && image.nodeName !== 'VIDEO')
		{
			var url = image.attr('data-imagefilter-src');
			//if (image.nodeName == 'VIDEO') //not a video, so don't need to check
			//	histogram = this.animatedHistograms[image.data('imagefilter-histogram-id')];
			//else
			histogram = this.histograms[url];
			if (!this.isPicture(image, histogram))
				apply = false;
		}
		image.toggleClass(filteredClass, apply);
	}
};

ImageFilterer.prototype.setCustomValue = function(key, value) {
	this.customValueCache[key] = value;
	if (this.defaultFilter)
		this.defaultFilter.setCustomValue(key, value);
	for (var k in this.filters)
		this.filters[k].setCustomValue(key, value);
};

ImageFilterer.prototype.setFilterSources = function(sources) {
	this.filterSources = sources;
	if (this.defaultFilter)
		this.defaultFilter.update(sources);
	for (var key in this.filters)
		this.filters[key].update(sources);
};

ImageFilterer.prototype.enable = function(enabled) {
	this.enabled = enabled;
	if (this.defaultFilter)
		this.defaultFilter.enable(enabled);
	for (var key in this.filters)
		this.filters[key].enable(enabled);
};

ImageFilterer.prototype.chooseFilter = function(img, histogram) {
	if (histogram.success)
	{
		if (!(histogram.id in this.filters))
			this.filters[histogram.id] = new ImageFilter(this.filterSources, this.enabled, this.customValueCache, this.inverted, histogram);
		return this.filters[histogram.id];
	}
	else
	{
		if (!this.defaultFilter)
			this.defaultFilter = new ImageFilter(this.filterSources, this.enabled, this.customValueCache, this.inverted);
		return this.defaultFilter;
	}
};

ImageFilterer.prototype.findHistogram = function(image) {
	var url = $(image).attr('data-imagefilter-src');
	if (this.nodeName == 'VIDEO')
		return this.animatedHistograms[$(image).data('imagefilter-histogram-id')];
	else
		return this.histograms[url];
}

//override a filter already applied to an image
ImageFilterer.prototype.applyManually = function(image, sources) {
	image = $(image);

	if (!$.inArray(image, this.images))
	{
		console.log('Cannot apply filter, image ' + image + " hasn\'t been added yet.");
		return;
	}

	var histogram = this.findHistogram(image);

	var filterID = image.data('imagefilter-override');
	if (filterID)
	{
		filter = this.filterOverrides[filterID];
		filter.release();
		delete this.filterOverrides[filterID];
	}
	else
	{
		filterID = this.nextOverrideID++;
		image.data('imagefilter-override', filterID);
	}

	if (sources === null)
	{
		image.removeData('imagefilter-override');
		this.applyFilterToImage([image], histogram);
	}
	else
	{
		var filter = new ImageFilter(sources, true, this.customValueCache, histogram);
		this.filterOverrides[filterID] = filter;

		//remove old css class
		image.removeClass(image.attr('data-imagefilter-class'));

		//add the new one
		image.addClass(filter.styleName);

		//store the class name as a backup
		image.attr('data-imagefilter-class', filter.styleName);
	}
};

ImageFilterer.prototype.applyFilterToImage = function(images, histogram) {
	var filteredImages = [];
	for (var i = 0; i < images.length; ++i)
	{
		var apply = true;
		if ($(images[i]).data('imagefilter-haschild'))
			apply = false;
		if (apply)
		{
			var filter = this.chooseFilter(images[i], histogram);

			var ancestors = $(images[i]).parents();
			ancestors.each(function(){
				//remove any filters applied to ancestors
				var filtered = $(this).attr('data-imagefilter-class');
				if (filtered)
					$(this).removeClass(filtered);

				//don't allow future filtering of ancestors
				$(this).data('imagefilter-haschild', ($(this).data('imagefilter-haschild') || 0) + 1);
			});

			//actually apply the filter by adding a class to the element
			if (!this.onlyPictures || images[i].nodeName == 'VIDEO' || this.isPicture(images[i], histogram))
				$(images[i]).addClass(filter.styleName);

			//store the class name as a backup, even if it wasn't applied
			$(images[i]).attr('data-imagefilter-class', filter.styleName);

			//NOTE: not used
			//apply the global class to mass allow filtering everything
			//$(images[i]).addClass('imagefilter-all');

			filteredImages.push(images[i]);
		}
	}
	$.merge(this.images, filteredImages);
};

ImageFilterer.prototype.updateFilter = function(histogram) {
	if (histogram.success)
	{
		this.filters[histogram.id].update();

		//update the debug histogram if it exists
		if (ImageFilterer.debugInfo && ImageFilterer.debugInfo.data('imagefilter-src') == histogram.src)
		{
			ImageFilterer.debugInfo.find('#imagefilter-histogram').replaceWith($(histogram.createGraph()).attr('id', 'imagefilter-histogram').css('border', '1px solid black'));
			ImageFilterer.debugInfo.find('#imagefilter-filterinfo').html(this.filters[histogram.id].getInfo());
		}
	}
};

ImageFilterer.prototype.removeFilter = function(histogram) {
	if (histogram.success)
		this.filters[histogram.id].remove();
};

ImageFilterer.prototype.sourceAdded = function(src, firstElement) {
	if (src && firstElement.nodeName != 'VIDEO')
	{
		console.log("Added " + src);
		this.histograms[src] = new Histogram(src, firstElement, 0);
		this.histograms[src].onload = this.histogramReady.bind(this);
		this.histograms[src].onerror = this.histogramReady.bind(this);
	}
};

ImageFilterer.prototype.sourceRemoved = function(src) {
	if (src in this.histograms)
	{
		console.log("Removed " + src);
		this.removeFilter(this.histograms[src]);
		this.histograms[src].stop();
		delete this.histograms[src];
	}
};

ImageFilterer.prototype.imageAdded = function(img, url) {
	if (img.nodeName == 'VIDEO')
	{
		var id = this.uidNext++;
		$(img).data('imagefilter-histogram-id', id);
		this.animatedHistograms[id] = new Histogram(url, img, this.animationUpdateFrequency);
		this.animatedHistograms[id].onload = this.histogramReady.bind(this);
		this.animatedHistograms[id].onupdate = this.updateFilter.bind(this);
	}
	else
	{
		if (url in this.histograms && this.histograms[url].ready)
			this.applyFilterToImage([img], this.histograms[url]);
	}

	var that = this;
	$(img).on('mouseover', function(e){
		e.stopPropagation();

		if (!ImageFilterer.enableDebug)
			return;

		//shouldn't need ImageFilterer.debugInfo.id == 'imagefilter-debug' but something weird is making it point to random elements
		if (ImageFilterer.debugInfo && ImageFilterer.debugInfo.data('imagefilter-stay'))
		{
			if (ImageFilterer.debugInfo.attr('id') == 'imagefilter-debug')
				return;
			else
				debugger;
		}
		if (ImageFilterer.timeout)
			window.clearTimeout(ImageFilterer.timeout);
		ImageFilterer.timeout = window.setTimeout(function(){
			$(ImageFilterer.debugInfo).css('pointer-events', 'auto').css('opacity', '1').data('imagefilter-stay', true)
		}, 2000);
		if (ImageFilterer.debugInfo)
			ImageFilterer.debugInfo.remove();
		ImageFilterer.debugInfo = $('<div id="imagefilter-debug" style="opacity: 0.5; pointer-events: none; z-index: 19999999999; position:fixed; top:0px; left:0px; right:0px; background: white; font-size: 14px; border-bottom: 2px solid black;" data-imagefilter-haschild="1"></div>');
		$(document.body).append(ImageFilterer.debugInfo);
		ImageFilterer.debugInfo.on('click', function(){
			ImageFilterer.debugInfo.remove();
			ImageFilterer.debugInfo = null;
		});

		var histogram;
		if (this.nodeName == 'VIDEO')
			histogram = that.animatedHistograms[$(this).data('imagefilter-histogram-id')];
		else
			histogram = that.histograms[url];

		var info = $('<div><div style="background:rgba(255,255,255,0.5);">URL: <a href="' + url + '">' + url + '</a></div></div>');
		ImageFilterer.debugInfo.append(info);

		if (histogram)
		{
			ImageFilterer.debugInfo.data('imagefilter-src', histogram.src);
			if (histogram.success)
			{
				ImageFilterer.debugInfo.append($(histogram.createGraph()).attr('id', 'imagefilter-histogram').css('border', '1px solid black'));
				info.css('position', 'absolute');
			}
			else
				info.append($('<div>Histogram status: '+histogram.status+'</div>'));
		}

		if (url && this.nodeName != 'VIDEO')
			ImageFilterer.debugInfo.append($('<img data-imagefilter-haschild="1" style="border:1px solid black; max-height: 128px;" alt="debugimg" src="' + url + '"/>'));

		var textInfo = $('<div style="display: inline-block" id="imagefilter-info"><div id="imagefilter-filterinfo"></div></div>');
		ImageFilterer.debugInfo.append(textInfo);
		if (this.nodeName != 'VIDEO')
		{
			var filteredClass = $(this).attr('data-imagefilter-class');
			textInfo.prepend('<div>Class Applied: ' + $(this).hasClass(filteredClass) + '</div>');
			textInfo.prepend('<div>Child Counter: ' + $(this).data('imagefilter-haschild') + '</div>');
			if (histogram.id in that.filters)
				textInfo.find('#imagefilter-filterinfo').html(that.filters[histogram.id].getInfo());
		}

	}).on('mouseout', function(){
		if (ImageFilterer.debugInfo && !ImageFilterer.debugInfo.data('imagefilter-stay'))
		{
			ImageFilterer.debugInfo.remove();
			ImageFilterer.debugInfo = null;
		}
		if (typeof ImageFilterer.timeout !== 'undefined')
		{
			window.clearTimeout(ImageFilterer.timeout);
			ImageFilterer.timeout = null;
		}
	});
};

ImageFilterer.prototype.imageRemoved = function(img, url, removedFrom) {
	if (img.nodeName == 'VIDEO')
	{
		var id = $(img).data('imagefilter-histogram-id');
		this.removeFilter(this.animatedHistograms[id]);
		this.animatedHistograms[id].stop();
		delete this.animatedHistograms[id];
	}

	//remove css class that applies the filter
	$(img).removeClass($(img).attr('data-imagefilter-class'));
	$(img).removeAttr('data-imagefilter-class');

	//remove image from list of filtered images if it exists
	var index = $.inArray(img, this.images);
	if (index)
		this.images.splice(index, 1);

	var ancestors = $(removedFrom).parents().addBack();

	//remove data-imagefilter-haschild contributions
	ancestors.each(function(){
		var val = $(this).data('imagefilter-haschild');
		if (val > 1)
			$(this).data('imagefilter-haschild', val-1);
		else
			$(this).removeData('imagefilter-haschild');
	});

	//check if an ancestor can now be filtered, since this has been removed
	//FIXME: not sure if this works as it's jquery data
	var next = $(removedFrom).closest('[data-imagefilter-class]');
	if (next.length && !next.data('imagefilter-haschild'))
	{
		var histogram = this.findHistogram(next);
		this.applyFilterToImage(next, histogram);
	}
};

ImageFilterer.prototype.histogramReady = function(histogram) {
	var images = this.finder.sources[histogram.src];
	this.applyFilterToImage(images, histogram);
};
